### PR TITLE
Move binder for dyn to each list item

### PR DESCRIPTION
**Pull request by jackh726**
This essentially changes `ty::Binder<&'tcx List<ExistentialTraitRef>>` to `&'tcx List<ty::Binder<ExistentialTraitRef>>`.

This is a first step in moving the `dyn Trait` representation closer to Chalk, which we've talked about in @rust-lang/wg-traits.

r? @nikomatsakis

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -87,7 +87,7 @@ pub struct CtxtInterners<'tcx> {
     substs: InternedSet<'tcx, InternalSubsts<'tcx>>,
     canonical_var_infos: InternedSet<'tcx, List<CanonicalVarInfo<'tcx>>>,
     region: InternedSet<'tcx, RegionKind>,
-    existential_predicates: InternedSet<'tcx, List<ExistentialPredicate<'tcx>>>,
+    poly_existential_predicates: InternedSet<'tcx, List<ty::Binder<ExistentialPredicate<'tcx>>>>,
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/context.rs:90]: This name change probably isn't necessary, but I would like to hear thoughts.

Comment (nikomatsakis) [reply compiler/rustc_middle/src/ty/context.rs:90]: seems fine

File: compiler/rustc_middle/src/ty/relate.rs [hunk 2]
```diff
@@ -616,24 +616,28 @@ impl<'tcx> Relate<'tcx> for &'tcx ty::List<ty::ExistentialPredicate<'tcx>> {
         // in `a`.
         let mut a_v: Vec<_> = a.into_iter().collect();
         let mut b_v: Vec<_> = b.into_iter().collect();
-        a_v.sort_by(|a, b| a.stable_cmp(tcx, b));
+        a_v.sort_by(|a, b| a.skip_binder().stable_cmp(tcx, &b.skip_binder()));
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/ty/relate.rs:619]: This seems fine, but I think it works only because we are moving towards a "normalized" form where bound things are referenced purely by index, and this binder has exactly one bound item.

Comment (nikomatsakis) [reply compiler/rustc_middle/src/ty/relate.rs:619]: Might be worth adding a comment about this.

Comment (nikomatsakis) [reply compiler/rustc_middle/src/ty/relate.rs:619]: (In particular, if we had something like `for<'a>...` and `for<'b>...`, and the late-bound regions referenced them by name, those are equivalent, but they might be sorted differently.)

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
r=me with comment

**Comment by jackh726** [comment]
Alright, so the trivial explanation here is simply that `stable_cmp` doesn't take into account anything other than the type of predicates. In a perfect world, I would go into more depth about the scenarios in which it *did*, but this actually is a somewhat complicated issue (see Zulip discussion [here](https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/Existential.20trait.20ref.20in.20rustc/near/219871899)), and it's not entirely clear that there could be multiple different bound vars here anyways.

Given that, @bors r=nikomatsakis

**Comment by jackh726** [comment]
@bors r-

NLL test needs to be blessed (this is causing the rollups to fail)

File: src/test/ui/issues/issue-40000.nll.stderr [hunk 3]
```diff
@@ -4,5 +4,11 @@ error: higher-ranked subtype error
 LL |     foo(bar);
    |         ^^^
 
-error: aborting due to previous error
+error: higher-ranked subtype error
+  --> $DIR/issue-40000.rs:6:9
+   |
+LL |     foo(bar);
+   |         ^^^
+
```
Comment (jackh726) [inline src/test/ui/issues/issue-40000.nll.stderr:12]: Not exactly sure why this particular error gets duplicated. But, this is a pretty unhelpful error anyways and other tests have similar duplication for it (sometimes multiple). So, any fix for this duplication should probably be part of a larger, unrelated fix.

**Comment by jackh726** [comment]
Test has been blessed. As I said in comment above, this particular error, while duplicated, isn't helpful, is duplicated in other tests, and a fix is an orthogonal change.

@bors r=nikomatsakis
